README.md
# Wos下载机器人

基于油猴脚本的wos核心期刊批量下载脚本，经过测试，可以一次性下载10万条数据。

安装地址：https://greasyfork.org/zh-CN/scripts/448810-wos-download-bot

## 功能说明

- 批量下载wos核心论文集全部数据
- 以条数重命名下载的文件
- 可以下载的格式：ris、bibtex、excel、txt
- 选择起始下载份数，不小心中断可选择中断时的份数继续下载

安装脚本后，进入wos，通过按期刊标题搜索，进入结果页，然后刷新一下，即可在右上角看到「一键下载」按钮，点击后按提示输入下载选项，最后等待下载完成即可。

![usage](images/usage.png)

输入下载选项

![usage](images/dialog.png)

以一定时间间隔，从wos数据库下载wos核心数据集，每次下载数量为500条，直到下载完或出错为止。

## 使用步骤

1、浏览器安装油猴插件

2、访问[greasyfork](https://greasyfork.org/zh-CN/scripts/448810-wos-download-bot)安装脚本。

3、登录wos。

4、按下面规则进行检索，进入检索结果页面。注意是**核心合集**，按出版物标题检索，页面不正确不会显示下面的按钮。

![检索页](images/search.png)

5、在右上角会出现「一键下载」的按钮，点击下载即可自动下载该期刊的所有论文。如果没出现该按钮，需要**刷新**一下当前页面。

![一键下载按钮](images/main.png)

6、输入配置，可以选择需要下载的文件类型。

7、下载结束后会出现「下载完成」的提示。

![下载完成](images/finish.png)

## 注意事项

- 下载时候不要关闭浏览器；
- 因为有环境检测，下载过程中建议时不时随意移动下鼠标，不要一直保持静止;
- 如果出现人机认证的要求导致无法下载，需要先手动下载一条数据，进行人机认证后，再重新「一键下载」;
- 使用谷歌浏览器可能会弹出请求下载多个文件的提示，点击确定即可;
- 由于服务器接口的限制，10万条后面的数据无法下载;

## 开发步骤

同使用步骤，之后

1、在tampermokey的编辑器修改代码，保存；
2、重新刷新相关url即可。

## 鸣谢

特别感谢[阳志平老师](https://github.com/ouyangzhiping)以及[开智学堂](https://github.com/OpenMindClub)信分十期


CHANGELOG
## v1.4.1

- 修复无法下载文件的bug

## v1.4.0

- 添加下载选项弹窗，包括文件格式，下载起始份数，下载速率
- 文件格式支持txt、ris、bibtex

## v1.3.1

- 更新获取`wosSid`变量的功能
- 更新获取`bm_telemetry`变量的功能

## v1.3.0

- 添加按给定起始序号下载论文的功能，增加输入框
- 重构下载功能
- 重构下载错误提示显示功能

## v1.2.4

- 修复下载的文件可能会丢失数据的问题


wos-batch-download.js
// ==UserScript==
// @name         Wos Batch Download Bot
// @namespace    http://tampermonkey.net/
// @version      1.2
// @description  一键下载wos核心合集的期刊论文数据
// @author       AngelLiang
// @match        https://www.webofscience.com/wos/woscc/summary/*/relevance/*
// @icon         https://www.google.com/s2/favicons?sz=64&domain=webofscience.com
// @require      https://cdn.staticfile.org/jquery/3.4.1/jquery.min.js
// @require      https://cdn.bootcss.com/jquery-cookie/1.4.1/jquery.cookie.js
// @connect      *
// ==/UserScript==


(function() {
    'use strict';

    const DOWNLOAD_URL = "https://www.webofscience.com/api/wosnx/indic/export/saveToFile";
    var total = "";
    var perPage = 500

    var uuid = ""
    var wosSid = ""
    var waitSecond = 60

    function getTotal() {
        total = $(".brand-blue").text()
        total = total.replace(",", "")
        total = parseInt(total)
        return total
    }

    // 根据index获取页码数量
    function getNextPage(index) {
        let start = 1 + index * perPage
        let stop = perPage + index * perPage
        return [""+start, ""+stop]
    }

    // 获取期刊的uuid
    function getParentQid() {
        let currUrl = String(window.location.href);
        return currUrl.split('/')[6]
    }

    // 生成请求页面数
    function genRequestPageNumber() {
        total = getTotal()
        let number = total / perPage
        return Math.ceil(number)
    }

    function get_bm_telemetry() {
        return bmak['sensor_data']
    }

    function genRequestData(start, stop) {
        let requestJson = {
            "parentQid": getParentQid(),
            "sortBy": "relevance",
            "displayTimesCited": "true",
            "displayCitedRefs": "true",
            "product": "UA",
            "colName": "WOS",
            "displayUsageInfo": "true",
            "fileOpt": "othersoftware",
            "action": "saveToFieldTagged",
            "markFrom": start,
            "markTo": stop,
            "view": "summary",
            "isRefQuery": "false",
            "locale": "en_US",
            "filters": "fullRecordPlus",
            "bm-telemetry": get_bm_telemetry()
        }
        let requestData = JSON.stringify(requestJson)
        return requestData
    }

    function downloadFile(fileName, data){
        var a = document.createElement("a");
        a.href = "data:text," + data;   //content
        a.download = fileName;            //file name
        a.click();
    }

    function sleep(delay) {
        for(var t = Date.now(); Date.now() - t <= delay;);
    }

    // 生成随机数
    function randomNum(minNum, maxNum){
        switch(arguments.length){
            case 1:
                return parseInt(Math.random()*minNum+1,10);
            case 2:
                return parseInt(Math.random()*(maxNum-minNum+1)+minNum,10);
            default:
                return 0;
        }
    }

    function enableDownloadButton(){
        $("#downloadButton").val('一键下载')
        $("#downloadButton").attr("disabled", false);
    }

    function disableDownloadButton(){
        $("#downloadButton").val('正在下载...')
        $("#downloadButton").attr("disabled", true);
    }

    function requestFile(i, number, total, callback) {
        if (waitSecond == null) {
            waitSecond = prompt("请输入下载间隔时间（±5s），单位秒", "60");
        }
        if (waitSecond ==  null) {
            alert("waitSecond参数不能为空");
            return
        }
        waitSecond = parseInt(waitSecond)
        if(waitSecond < 20 ) {
            alert("下载间隔时间不能小于20s");
            return
        }

        var nextPageParam = getNextPage(i - 1)
        var start = nextPageParam[0]
        var stop = nextPageParam[1]
        if (stop > total) {
            stop = "" + total;
        }
        //console.log(nextPageParam)

        var requestData = genRequestData(start, stop)
        console.log("正在下载" + start + "到" + stop + "份数据，总共" + total + "份")
        disableDownloadButton()

        let reqAjax = $.ajax({
            url:DOWNLOAD_URL,
            type: 'POST',
            headers: {
                "x-1p-wos-sid": wosSid,
                "content-type": "application/json, text/plain, */*",
                "accept-language": "zh-CN,zh;q=0.9"
            },
            "crossDomain": true,
            data: requestData,
            success: function(result){
                //console.log(result);
                let filename = ""+ nextPageParam[0] + '-' + nextPageParam[1] + '.txt'
                downloadFile(filename, result)
            },
            error: function(err){
                console.log(err);
                enableDownloadButton()
                alert('下载出错')
            }
        })

        $.when(reqAjax).done(function(){
            var nextIndex = i+1
            if (nextIndex > number) {
                enableDownloadButton()
                console.log("===下载完成===");
                alert('下载完成')
                return
            }
            var sleepNum = randomNum(waitSecond - 5, waitSecond + 5)
            console.log("等待" + sleepNum + "秒后再下载")
            sleep(sleepNum*1000)
            callback(nextIndex, number, total, callback)
        })
    }

    function getSessionID(doc) {
        const sidRegex = /sid=([a-zA-Z0-9]+)/i;

        // session ID is embedded in the static page inside an inline <script>
        // if you have the right HttpOnly cookie set. if we can't find it, we
        // initialize our session as the web app does
        for (let scriptTag of doc.querySelectorAll('script')) {
            let sid = scriptTag.textContent.match(sidRegex);
            if (sid) {
                return sid[1];
            }
        }
        return null
    }

    function downloadCallback() {
        console.log("===下载文件===")
        // wosSid = getSessionID(document)
        // console.log(wosSid)
        // uuid = getParentQid()
        // console.log(uuid)
        let total = getTotal()
        // console.log(total)
        let number = genRequestPageNumber(total)
        requestFile(1, number, total, requestFile)
    }

    function addButton() {
        console.log("addButton")
        $('body').append('<button id="downloadButton">一键下载</button>')
        $('#downloadButton').css('width', '120px')
        $('#downloadButton').css('position', 'absolute')
        $('#downloadButton').css('top', '120px')
        $('#downloadButton').css('right', '50px')
        $('#downloadButton').css('background-color', '#5e33bf')
        $('#downloadButton').css('color', 'white')
        $('#downloadButton').css('font-size', 'large')
        $('#downloadButton').css('z-index', 100)
        $('#downloadButton').css('border-radius', '5px')
        $('#downloadButton').css('text-align', 'center')

        $('#downloadButton').click(downloadCallback)
    };

    $(document).ready(function () {
        if (window.location.href.startsWith('https://www.webofscience.com')) {
            addButton()
        }
    })

})();


wos-download-bot.js
// ==UserScript==
// @name         Wos Download Bot
// @namespace    http://tampermonkey.net/
// @version      1.4.1
// @description  wos核心论文集下载机器人
// @author       AngelLiang
// @match        https://www.webofscience.com/wos/woscc/summary/*/relevance/*
// @icon         https://www.google.com/s2/favicons?sz=64&domain=webofscience.com
// @require      https://cdn.staticfile.org/jquery/3.4.1/jquery.min.js
// @require      https://cdn.bootcss.com/jquery-cookie/1.4.1/jquery.cookie.js
// @require      https://code.jquery.com/ui/1.13.0/jquery-ui.min.js
// @grant        GM_download
// @connect      *
// @license      MIT
// ==/UserScript==


(function() {
    'use strict';

    const DOWNLOAD_URL = "https://www.webofscience.com/api/wosnx/indic/export/saveToFile";

    var total = "";
    // 论文集的唯一码
    var uuid = ""
    // wos的唯一码（？）
    var wosSid = ""
    // 等待秒数
    var waitSecond = 30
    var fileFormat = ''

    function getTotal() {
        total = $(".brand-blue").text()
        total = total.replace(",", "")
        total = parseInt(total)
        return total
    }

    // 获取期刊的uuid
    function getParentQid() {
        let currUrl = String(window.location.href);
        return currUrl.split('/')[6]
    }

    function get_bm_telemetry() {
        // return bmak['sensor_data']
        // update: 20230709修改接口
        return bmak.get_telemetry()
    }

    function get_fileOpt() {
        if (fileFormat == 'excel') {
            return "xls"
        } else if (fileFormat == 'ris') {
            return "othersoftware"
        } else if (fileFormat == 'bibtex'){
            return "othersoftware"
        } else {
            return "othersoftware"
        }
    }

    function get_action() {
        if (fileFormat == 'excel') {
            return "saveToExcel"
        } else if (fileFormat == 'ris') {
            return "saveToRIS"
        } else if (fileFormat == 'bibtex'){
            return "saveToBibtex"
        } else {
            return "saveToFieldTagged"
        } 
    }

    function get_filters() {
        if (fileFormat == 'excel') {
            return "fullRecord"
        } else if (fileFormat == 'ris') {
            return "fullRecord"
        } else if (fileFormat == 'bibtex'){
            return "fullRecordPlus"
        } else {
            return "fullRecordPlus"
        } 
    }

    function getDownloadFileExt() {
        if (fileFormat == 'excel') {
            return ".xls"
        } else if (fileFormat == 'ris') {
            return ".ris"
        } else if (fileFormat == 'bibtex'){
            return ".bib"
        } else {
            return ".txt"
        } 
    }

    function genRequestData(start, stop) {
        let requestJson = {
            "parentQid": getParentQid(),
            "sortBy": "relevance",
            "displayTimesCited": "true",
            "displayCitedRefs": "true",
            "product": "UA",
            "colName": "WOS",
            "displayUsageInfo": "true",
            "fileOpt": get_fileOpt(),
            "action": get_action(),
            "markFrom": start,
            "markTo": stop,
            "view": "summary",
            "isRefQuery": "false",
            "locale": "en_US",
            "filters": get_filters(),
            // update: 20231211 不再需要这个字段
            // "bm-telemetry": get_bm_telemetry()
        }
        let requestData = JSON.stringify(requestJson)
        return requestData
    }

    function downloadFile(fileName, content) {
        let downLink = document.createElement('a')
        downLink.download = fileName
        //字符内容转换为blod地址
        let blob = new Blob([content])
        downLink.href = URL.createObjectURL(blob)
        // 链接插入到页面
        document.body.appendChild(downLink)
        downLink.click()
        // 移除下载链接
        document.body.removeChild(downLink)
    }

    function sleep(delay) {
        for(var t = Date.now(); Date.now() - t <= delay;);
    }

    // 生成随机数
    function randomNum(minNum, maxNum){
        switch(arguments.length){
            case 1:
                return parseInt(Math.random()*minNum+1,10);
            case 2:
                return parseInt(Math.random()*(maxNum-minNum+1)+minNum,10);
            default:
                return 0;
        }
    }

    function enableDownloadButton(){
        $("#downloadButton").html('一键下载')
        $("#downloadButton").attr("disabled", false);
    }

    function disableDownloadButton(){
        $("#downloadButton").html('正在下载...')
        $("#downloadButton").attr("disabled", true);
    }

    function requestFileByIndex(start, stop, total, callback) {
        start = "" + start;
        if (parseInt(stop) > total) {
            stop = total;
        }
        stop = "" + stop;

        console.log("正在下载" + start + "到" + stop + "份数据，总共" + total + "份")
        disableDownloadButton()
        try {
            var data = genRequestData(start, stop)    
        } catch (error) {
            console.error(error)
            enableDownloadButton()
            alert('请求出错，请联系开发者')
            return
        }

        let reqAjax = $.ajax({
            url:DOWNLOAD_URL,
            type: 'POST',
            headers: {
                "x-1p-wos-sid": wosSid,
                "accept": "application/json, text/plain, */*",
                "content-type": "application/json",
                "accept-language": "zh-CN,zh;q=0.9"
            },
            "crossDomain": true,
            data: data,
            success: function(result){
                //console.log(result);
                let filename = "" + start + '-' + stop + getDownloadFileExt()
                downloadFile(filename, result)
            },
            error: function(err){
                console.log(err);
                console.log(err.responseJSON);
                enableDownloadButton()
                alert('下载出错：'+JSON.stringify(err.responseJSON))
            }
        })

        $.when(reqAjax).done(function(){
            var nextStart = parseInt(start) + 500
            var nextStop = parseInt(stop) + 500
            if (nextStart > total) {
                enableDownloadButton()
                console.log("===下载完成===");
                alert('下载完成')
                return
            }
            let min_val = waitSecond - 5
            let max_val = waitSecond + 5
            var sleepNum = randomNum(min_val, max_val)
            console.log("等待" + sleepNum + "秒后再下载")
            sleep(sleepNum*1000)
            callback(nextStart, nextStop, total, callback)
        })
    }

    function getSessionID() {
        // const sidRegex = /SID=([a-zA-Z0-9]+)/i;

        // // session ID is embedded in the static page inside an inline <script>
        // // if you have the right HttpOnly cookie set. if we can't find it, we
        // // initialize our session as the web app does
        // for (let scriptTag of document.querySelectorAll('script')) {
        //     let sid = scriptTag.textContent.match(sidRegex);
        //     if (sid) {
        //         return sid[1];
        //     }
        // }
        // return null

        return sessionData.BasicProperties.SID
    }

    // 创建弹窗的函数
    function createPopup() {
        // 创建弹窗的容器
        var popupContainer = document.createElement('div');
        popupContainer.style.position = 'fixed';
        popupContainer.style.top = '50%';
        popupContainer.style.left = '50%';
        popupContainer.style.transform = 'translate(-50%, -50%)';
        popupContainer.style.background = '#fff';
        popupContainer.style.padding = '20px';
        popupContainer.style.border = '1px solid #ccc';
        popupContainer.style.zIndex = '9999';
    
        // 添加设置选项到弹窗
        // bug: 下载的excel还有点问题
        popupContainer.innerHTML = `
        <h3>一键下载</h3>
        <label for="fileFormat">文件格式：</label>
        <select id="fileFormat">
            <option value="ris">RIS</option>
            <option value="bibtex">BibTeX</option>
            <option value="txt">txt</option>
        </select><br>
        <label for="startDownloadFrom">起始下载份数：</label>
        <input type="text" id="startDownloadFrom" value="1"><br>
        <label for="downloadSpeed">下载速率（单位秒）：</label>
        <input type="text" id="downloadSpeed" value="30"><br>
        <button id="confirmButton">开始下载</button>
        <button id="cancelButton">取消</button>
        `;
    
        // 确定下载按钮的点击事件处理
        var confirmButton = popupContainer.querySelector('#confirmButton');
        confirmButton.addEventListener('click', function() {
            waitSecond = parseInt(document.querySelector('#downloadSpeed').value);
            fileFormat = document.querySelector('#fileFormat').value;
            var start = parseInt(document.querySelector('#startDownloadFrom').value);

            // 进行下载操作
            console.log("开始下载")
            wosSid = getSessionID()
            uuid = getParentQid()
            var total = getTotal()

            console.log('下载速率：', waitSecond)
            console.log('文件格式:', fileFormat)
            console.log('起始下载份数：', start)
            console.log('wosSid:', wosSid)
            console.log('uuid:', uuid)
            console.log('总数：', total)

            var stop = start + 500 - 1

            // 校验
            if (waitSecond < 20){
                alert('下载速率不能太快')
                return
            }
            if (start > total) {
                alert('超过下载总数')
                return
            }

            requestFileByIndex(start, stop, total, requestFileByIndex)
            // 移除弹窗
            document.body.removeChild(popupContainer);
        });
    
        // 取消按钮的点击事件处理
        var cancelButton = popupContainer.querySelector('#cancelButton');
            cancelButton.addEventListener('click', function() {
            console.log("取消下载")
            // 移除弹窗
            document.body.removeChild(popupContainer);
        });
    
        // 将弹窗添加到页面
        document.body.appendChild(popupContainer);
    }

    function downloadCallback() {
        createPopup();
    }

    function initButton() {
        $('body').append('<button id="downloadButton">一键下载</button>')
        $('#downloadButton').css('width', '120px')
        $('#downloadButton').css('position', 'absolute')
        $('#downloadButton').css('top', '120px')
        $('#downloadButton').css('right', '50px')
        $('#downloadButton').css('background-color', '#5e33bf')
        $('#downloadButton').css('color', 'white')
        $('#downloadButton').css('font-size', 'large')
        $('#downloadButton').css('z-index', 100)
        $('#downloadButton').css('border-radius', '5px')
        $('#downloadButton').css('text-align', 'center')

        $('#downloadButton').click(downloadCallback)

        // 将按钮添加到页面
        // document.body.appendChild(downloadButton);
    };

    // 检查URL并初始化按钮
    function checkAndInitButton() {
        if (window.location.href.includes('https://www.webofscience.com/')) {
            initButton();
        }
    }

    $(document).ready(function () {
        console.log("WDB-v1.4.1")
        checkAndInitButton();
    })

})();
